### PR TITLE
TypeScript migration

### PR DESCRIPTION
**Pull request by jonbarrow**
Creating as a draft PR because a lot more work and testing needs to be done to ensure everything still works. This is the first step towards moving the account server to TypeScript

Tested working features:

- [x] Create PNID
- [x] Link existing PNID
- [x] Login to linked PNID
- [x] Remove user/unlink PNID
- [x] Connecting to game servers
- [x] Update Mii
- [x] Send emails
- [x] Verify email code
- [x] Website integration (likely needs more tasks)
- [x] 3DS integration

### CODE REVIEW

**Comment by jonbarrow** [comment]
I believe I have checked all major functionality and confirmed it is working. There is still more which needs to be done, however, with regards to getting the type system up to par. At the moment there is very little typing being done, and it's an akward mix of typed and un-typed data

**Comment by jonbarrow** [comment]
@mrjvs @Pandawan if you wouldn't mind giving this a look-over when you have time please

**Review by mrjvs — CHANGES_REQUESTED** [review]
I skipped most the actual logic, I only really looked at the use of typescript. (as that is what is being changed in the PR)

File: src/cache.ts [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+	getServicePrivateKey,
+	getServiceSecretKey,
+	getServiceAESKey,
+	setServicePublicKey,
+	setServicePrivateKey,
+	setServiceSecretKey,
+	setServiceAESKey,
+	getLocalCDNFile,
+	setLocalCDNFile
+};
```
[the remark this answers is not in the document]
Comment (jonbarrow) [reply src/cache.ts:213]: This was largely leftover from the require->import part of the migration, and I found that replacing `module.exports` with `export default` resulted in less broken imports at the time. I agree this is messy though, thank you for catching this!

File: src/database.ts [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+		throw new Error('Cannot make database requets without being connected');
+	}
+}
+
+async function getUserByUsername(username: string): Promise<HydratedPNIDDocument> {
+	verifyConnected();
+
+	if (typeof username !== 'string') {
+		return null;
+	}
```
[the remark this answers is not in the document]
Comment (jonbarrow) [reply src/database.ts:50]: Left over from before TS, good catch!

File: src/database.ts [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+	return await PNID.findOne({
+		pid
+	}) as HydratedPNIDDocument;
+}
+
+async function getUserByEmailAddress(email: string): Promise<HydratedPNIDDocument> {
+	verifyConnected();
+
+	return await PNID.findOne({
+		'email.address': new RegExp(email, 'i') // * Ignore case
```
[the remark this answers is not in the document]
Comment (jonbarrow) [reply src/database.ts:69]: I'm confused as to why this is more unsafe than storing the email twice? Storing the email in lowercase and then converting the input to lowercase here would achive the same result as just ignoring the case on the query, no? And if it's not a valid email address it should just not return a document, which is expected, is it not?

File: src/database.ts [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+	return !!await getUserByUsername(username);
+}
+
+async function getUserBasic(token: string): Promise<HydratedPNIDDocument> {
+	verifyConnected();
+
+	// * Wii U sends Basic auth as `username password`, where the password may not have spaces
+	// * This is not to spec, but that is the consoles fault not ours
+	const decoded: string = Buffer.from(token, 'base64').toString();
+	const parts: string[] = decoded.split(' ');
```
[the remark this answers is not in the document]
Comment (jonbarrow) [reply src/database.ts:85]: Putting this here as well for completeness sake even though we discussed this on Discord: After discussing this further this is really a matter of opinion/preference and mine is to fully use the type system

File: src/database.ts [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+		device_attributes = device.get('device_attributes').map(({name, value, created_date}) => ({
+			device_attribute: {
+				name,
+				value,
+				created_date: created_date ? created_date : ''
+			}
+		}));
+	}
+
+	return <PNIDProfile>{
```
[the remark this answers is not in the document]
Comment (jonbarrow) [reply src/database.ts:153]: Is this cast unneccesary when viewed from the eyes of "let TS infer the type"? If so, does this comment still apply now that you know I prefer to strictly type everything?

[the remark this answers is not in the document]
Comment (mrjvs) [reply src/database.ts:153]: yes it does still apply, cuz casting is fundementally lying to typescript, you're telling the compiler that yes, this is actually this type that im telling it to be, even if its not. so you could be returning data thats missing a bunch of fields and typescript would not complain.

for example, with this cast, typescript will not complain at all even tho its completely wrong:


but with strictly annotated types, ts does complain:

File: src/database.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mrjvs) [reply src/database.ts:69]: ill dm you about it

**Comment by jonbarrow** [comment]
@mrjvs I believe I have resolved all issues, look good?

**Pull request merged**